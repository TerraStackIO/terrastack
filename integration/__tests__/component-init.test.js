const fs = require("fs-extra");
const os = require("os");
const path = require("path");
const { runTerrastack } = require("../runTerrastack");

describe("Component Init", () => {
  const dir = path.resolve(__dirname, "..", "component-init");
  const workDir = path.resolve(os.tmpdir(), "component-init");

  beforeEach(() => {
    fs.removeSync(path.resolve(workDir, "*"));
    fs.copySync(dir, workDir);
  });

  it("generates a Terrastack index.js file", () => {
    const result = runTerrastack(workDir, [
      "component",
      "init",
      "-t",
      "1.0.0",
      "-d",
      "description",
      "foo-module"
    ]);

    const generatedIndexContent = fs
      .readFileSync(
        path.resolve(workDir, ".terrastack", "component", "index.js")
      )
      .toString();

    expect(generatedIndexContent).toMatchSnapshot();
  });
});
