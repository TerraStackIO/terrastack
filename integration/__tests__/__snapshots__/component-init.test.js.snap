// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Init generates a Terrastack index.js file 1`] = `
"const path = require(\\"path\\");

/**
* @typedef InputProperties
* @type {object}
* @property {string} foo
* @property {string} foo_with_description - bar.
* @property {Array} foo_list
* @property {Object.<string, (string|number)>} foo_map
* @property {string} [foo_with_default_string=\\"bar\\"] - Default: \\"bar\\"
* @property {Array} [foo_default_list=[]] - Default: []
* @property {Object.<string, (string|number)>} [foo_default_map={}] - Default: {}
* @property {boolean} [foo_default_boolean=true] - Default: true
* @property {number} [foo_default_integer=1] - Default: 1
* @property {string} [foo_with_default_string_and_description=\\"bar\\"] - foo_bar. Default: \\"bar\\"
*/

/**
 * inputCallback to configure the component input.
 * @callback inputCallback
 * @param {object} bindings
 * @returns {InputProperties} inputs object for the component
 */

class Component {

  /**
   * Initialize the component
   * @param {string} name - Unique name for the component
   * @param {object} bindings - Define dependencies for this component. Will be passed as argument to the optionsCallback
   * @param {inputCallback} inputCallback - Configure the component via the return value of this callback
   */
  constructor(name, bindings, inputCallback) {
    this.name = name;
    this.inputCallback = inputCallback;
    this.bindings = bindings;
    this.sourceDir = path.join(__dirname, \\"../..\\");
    this.version = 1
  }
}

module.exports = Component;
"
`;
